File: iMEGA/Camera uploads/UploadUtils/CameraUploadConcurrentCountCalculator.m
#import "CameraUploadConcurrentCountCalculator.h"

@interface CameraUploadConcurrentCountCalculator ()

@property (nonatomic) CameraUploadConcurrentCounts currentConcurrentCounts;

@end

@implementation CameraUploadConcurrentCountCalculator

#pragma mark - notifications to monitor

- (void)startCalculatingConcurrentCount {
    [NSNotificationCenter.defaultCenter removeObserver:self];
    
    [NSNotificationCenter.defaultCenter addObserver:self selector:@selector(applicationStatesChangedNotification:) name:UIApplicationDidEnterBackgroundNotification object:nil];
    [NSNotificationCenter.defaultCenter addObserver:self selector:@selector(applicationStatesChangedNotification:) name:UIApplicationDidBecomeActiveNotification object:nil];
    
    [NSNotificationCenter.defaultCenter addObserver:self selector:@selector(applicationStatesChangedNotification:) name:UIDeviceBatteryStateDidChangeNotification object:nil];
    [NSNotificationCenter.defaultCenter addObserver:self selector:@selector(applicationStatesChangedNotification:) name:UIDeviceBatteryLevelDidChangeNotification object:nil];
    [NSNotificationCenter.defaultCenter addObserver:self selector:@selector(applicationStatesChangedNotification:) name:NSProcessInfoPowerStateDidChangeNotification object:nil];
    
    [NSNotificationCenter.defaultCenter addObserver:self selector:@selector(applicationStatesChangedNotification:) name:NSProcessInfoThermalStateDidChangeNotification object:nil];
}

- (void)stopCalculatingConcurrentCount {
    [NSNotificationCenter.defaultCenter removeObserver:self];
}

- (void)applicationStatesChangedNotification:(NSNotification *)notification {
    MEGALogDebug(@"[Camera Upload] concurrent calculator received %@", notification.name);
    CameraUploadConcurrentCounts concurrentCounts = [self calculateCameraUploadConcurrentCounts];
    if (concurrentCounts.photoConcurrentCount != self.currentConcurrentCounts.photoConcurrentCount) {
        [NSNotificationCenter.defaultCenter postNotificationName:MEGACameraUploadPhotoConcurrentCountChangedNotification object:self userInfo:@{MEGAPhotoConcurrentCountUserInfoKey : @(concurrentCounts.photoConcurrentCount)}];
    }
    
    if (concurrentCounts.videoConcurrentCount != self.currentConcurrentCounts.videoConcurrentCount) {
        [NSNotificationCenter.defaultCenter postNotificationName:MEGACameraUploadVideoConcurrentCountChangedNotification object:self userInfo:@{MEGAVideoConcurrentCountUserInfoKey : @(concurrentCounts.videoConcurrentCount)}];
    }
    
    self.currentConcurrentCounts = concurrentCounts;
}

#pragma mark - concurrent count calculation

- (PhotoUploadConcurrentCount)calculatePhotoUploadConcurrentCount {
    self.currentConcurrentCounts = [self calculateCameraUploadConcurrentCounts];
    return self.currentConcurrentCounts.photoConcurrentCount;
}

- (VideoUploadConcurrentCount)calculateVideoUploadConcurrentCount {
    self.currentConcurrentCounts = [self calculateCameraUploadConcurrentCounts];
    return self.currentConcurrentCounts.videoConcurrentCount;
}

- (CameraUploadConcurrentCounts)calculateCameraUploadConcurrentCounts {
    if (NSThread.isMainThread) {
        return [self calculateCameraUploadConcurrentCountsInMainThread];
    } else {
        __block CameraUploadConcurrentCounts counts;
        dispatch_sync(dispatch_get_main_queue(), ^{
            counts = [self calculateCameraUploadConcurrentCountsInMainThread];
        });
        return counts;
    }
}

- (CameraUploadConcurrentCounts)calculateCameraUploadConcurrentCountsInMainThread {
    return [self calculateCameraUploadConcurrentCountsByThermalState:NSProcessInfo.processInfo.thermalState
                                                    applicationState:UIApplication.sharedApplication.applicationState
                                                        batteryState:UIDevice.currentDevice.batteryState
                                                        batteryLevel:UIDevice.currentDevice.batteryLevel
                                               isLowPowerModeEnabled:NSProcessInfo.processInfo.isLowPowerModeEnabled];
}

- (CameraUploadConcurrentCounts)calculateCameraUploadConcurrentCountsByThermalState:(NSProcessInfoThermalState)thermalState applicationState:(UIApplicationState)applicationState batteryState:(UIDeviceBatteryState)batteryState batteryLevel:(float)batteryLevel isLowPowerModeEnabled:(BOOL)isLowPowerModeEnabled {
    CameraUploadConcurrentCounts concurrentCounts = [self calculateCameraUploadConcurrentCountsByApplicationState:applicationState batteryState:batteryState batteryLevel:batteryLevel isLowPowerModeEnabled:isLowPowerModeEnabled];
    CameraUploadConcurrentCounts countsByThermal = [self concurrentCountsByThermalState:thermalState];
    return MakeCounts(MIN(concurrentCounts.photoConcurrentCount, countsByThermal.photoConcurrentCount), MIN(concurrentCounts.videoConcurrentCount, countsByThermal.videoConcurrentCount));
}

- (CameraUploadConcurrentCounts)calculateCameraUploadConcurrentCountsByApplicationState:(UIApplicationState)applicationState batteryState:(UIDeviceBatteryState)batteryState batteryLevel:(float)batteryLevel isLowPowerModeEnabled:(BOOL)isLowPowerModeEnabled {
    CameraUploadConcurrentCounts countsByAppState = [self concurrentCountsByApplicationState:applicationState];
    CameraUploadConcurrentCounts countsByPower = [self concurrentCountsByBatteryState:batteryState batteryLevel:batteryLevel isLowPowerModeEnabled:isLowPowerModeEnabled];
    return MakeCounts(MIN(countsByAppState.photoConcurrentCount, countsByPower.photoConcurrentCount), MIN(countsByAppState.videoConcurrentCount, countsByPower.videoConcurrentCount));
}

- (CameraUploadConcurrentCounts)concurrentCountsByApplicationState:(UIApplicationState)applicationState {
    if (applicationState == UIApplicationStateBackground) {
        return MakeCounts(PhotoUploadConcurrentCountInBackground, VideoUploadConcurrentCountInBackground);
    } else {
        return MakeCounts(PhotoUploadConcurrentCountInDefaultMaximum, VideoUploadConcurrentCountInDefaultMaximum);
    }
}

- (CameraUploadConcurrentCounts)concurrentCountsByBatteryState:(UIDeviceBatteryState)batteryState batteryLevel:(float)batteryLevel isLowPowerModeEnabled:(BOOL)isLowPowerModeEnabled {
    if (batteryState == UIDeviceBatteryStateUnplugged) {
        if (batteryLevel < 0.15) {
            return MakeCounts(PhotoUploadConcurrentCountInBatteryLevelBelow15, VideoUploadConcurrentCountInBatteryLevelBelow15);
        } else if (batteryLevel < 0.25) {
            return MakeCounts(PhotoUploadConcurrentCountInBatteryLevelBelow25, VideoUploadConcurrentCountInBatteryLevelBelow25);
        } else if (isLowPowerModeEnabled) {
            return MakeCounts(PhotoUploadConcurrentCountInLowPowerMode, VideoUploadConcurrentCountInLowPowerMode);
        } else if (batteryLevel < 0.4) {
            return MakeCounts(PhotoUploadConcurrentCountInBatteryLevelBelow40, VideoUploadConcurrentCountInBatteryLevelBelow40);
        } else if (batteryLevel < 0.55) {
            return MakeCounts(PhotoUploadConcurrentCountInBatteryLevelBelow55, VideoUploadConcurrentCountInBatteryLevelBelow55);
        } else if (batteryLevel < 0.75) {
            return MakeCounts(PhotoUploadConcurrentCountInBatteryLevelBelow75, VideoUploadConcurrentCountInBatteryLevelBelow75);
        } else {
            return MakeCounts(PhotoUploadConcurrentCountInBatteryLevel75OrAbove, VideoUploadConcurrentCountInBatteryLevel75OrAbove);
        }
    } else {
        return MakeCounts(PhotoUploadConcurrentCountInBatteryCharging, VideoUploadConcurrentCountInBatteryCharging);
    }
}

- (CameraUploadConcurrentCounts)concurrentCountsByThermalState:(NSProcessInfoThermalState)thermalState {
    switch (thermalState) {
        case NSProcessInfoThermalStateCritical:
            return MakeCounts(PhotoUploadConcurrentCountInThermalStateCritical, VideoUploadConcurrentCountInThermalStateCritical);
        case NSProcessInfoThermalStateSerious:
            return MakeCounts(PhotoUploadConcurrentCountInThermalStateSerious, VideoUploadConcurrentCountInThermalStateSerious);
        case NSProcessInfoThermalStateFair:
            return MakeCounts(PhotoUploadConcurrentCountInThermalStateFair, VideoUploadConcurrentCountInThermalStateFair);
        case NSProcessInfoThermalStateNominal:
            return MakeCounts(PhotoUploadConcurrentCountInDefaultMaximum, VideoUploadConcurrentCountInDefaultMaximum);
    }
}

@end

File: iMEGA/Camera uploads/UploadUtils/CameraUploadConcurrentCountCalculator.swift
import Foundation
import UIKit

enum CameraUploadMediaTypePausedReason: Sendable, Equatable {
    enum ThermalState: Sendable, Equatable {
        case critical
        case serious
    }
    case lowBattery
    case thermalState(ThermalState)
}

@objc class CameraUploadConcurrentCountCalculator: NSObject {
    private var currentUploadQueueStates = CameraUploadQueueStates(
            photoUploadState: .defaultMaximum,
            videoUploadState: .defaultMaximum)
    
    // MARK: - Notifications to monitor
    
    @objc func startCalculatingConcurrentCount() {
        NotificationCenter.default.removeObserver(self)
        
        NotificationCenter.default.addObserver(self, selector: #selector(applicationStatesChangedNotification(_:)), name: UIApplication.didEnterBackgroundNotification, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(applicationStatesChangedNotification(_:)), name: UIApplication.didBecomeActiveNotification, object: nil)
        
        NotificationCenter.default.addObserver(self, selector: #selector(applicationStatesChangedNotification(_:)), name: UIDevice.batteryStateDidChangeNotification, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(applicationStatesChangedNotification(_:)), name: UIDevice.batteryLevelDidChangeNotification, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(applicationStatesChangedNotification(_:)), name: Notification.Name.NSProcessInfoPowerStateDidChange, object: nil)
        
        NotificationCenter.default.addObserver(self, selector: #selector(applicationStatesChangedNotification(_:)), name: ProcessInfo.thermalStateDidChangeNotification, object: nil)
    }
    
    @objc func stopCalculatingConcurrentCount() {
        NotificationCenter.default.removeObserver(self)
    }
    
    @objc private func applicationStatesChangedNotification(_ notification: Notification) {
        MEGALogDebug("[Camera Upload] concurrent calculator received \(notification.name)")
        Task { @MainActor in
            let uploadQueueState = calculateCameraUploadQueueStates()
            
            if currentUploadQueueStates.photoConcurrentCount != uploadQueueState.photoConcurrentCount {
                var userInfo: [AnyHashable: Any] = [MEGAPhotoConcurrentCountUserInfoKey: uploadQueueState.photoConcurrentCount]
                if let pausedReason = uploadQueueState.photoPausedReason {
                    userInfo[MEGACameraUploadsPhotosPausedReasonUserInfoKey] = pausedReason
                }
                NotificationCenter.default.post(
                    name: .MEGACameraUploadPhotoConcurrentCountChanged,
                    object: self,
                    userInfo: userInfo)
            }
            
            if currentUploadQueueStates.videoConcurrentCount != uploadQueueState.videoConcurrentCount {
                var userInfo: [AnyHashable: Any] = [MEGAVideoConcurrentCountUserInfoKey: uploadQueueState.videoConcurrentCount]
                if let pausedReason = uploadQueueState.videoPausedReason {
                    userInfo[MEGACameraUploadsVideosPausedReasonUserInfoKey] = pausedReason
                }
                NotificationCenter.default.post(
                    name: .MEGACameraUploadVideoConcurrentCountChanged,
                    object: self,
                    userInfo: userInfo)
            }
            
            currentUploadQueueStates = uploadQueueState
        }
    }
    
    // MARK: - Concurrent count calculation
    
    @MainActor
    @objc func calculatePhotoUploadConcurrentCount() -> Int {
        currentUploadQueueStates = calculateCameraUploadQueueStates()
        return currentUploadQueueStates.photoConcurrentCount
    }
    
    @MainActor
    @objc func calculateVideoUploadConcurrentCount() -> Int {
        currentUploadQueueStates = calculateCameraUploadQueueStates()
        return currentUploadQueueStates.videoConcurrentCount
    }
    
    func photoQueuePausedReason() -> CameraUploadMediaTypePausedReason? {
        currentUploadQueueStates.photoPausedReason
    }
    
    func videoQueuePausedReason() -> CameraUploadMediaTypePausedReason? {
        currentUploadQueueStates.videoPausedReason
    }
    
    @MainActor
    private func calculateCameraUploadQueueStates() -> CameraUploadQueueStates {
        let statuses = [
            queueStatusByApplicationState(UIApplication.shared.applicationState),
            queueStatusByBatteryState(
                UIDevice.current.batteryState,
                batteryLevel: UIDevice.current.batteryLevel,
                isLowPowerModeEnabled: ProcessInfo.processInfo.isLowPowerModeEnabled),
            queueStatusByThermalState(ProcessInfo.processInfo.thermalState)
        ]
        return CameraUploadQueueStates(
            photoUploadState: statuses.min(by: { $0.photoConcurrentCount < $1.photoConcurrentCount }) ?? .defaultMaximum,
            videoUploadState: statuses.min(by: { $0.videoConcurrentCount < $1.videoConcurrentCount }) ?? .defaultMaximum)
    }
    
    private func queueStatusByApplicationState(_ applicationState: UIApplication.State) -> CameraUploadQueueState {
        if applicationState == .background {
            .background
        } else {
            .defaultMaximum
        }
    }
    
    private func queueStatusByBatteryState(_ batteryState: UIDevice.BatteryState,
                                           batteryLevel: Float,
                                           isLowPowerModeEnabled: Bool
    ) -> CameraUploadQueueState {
        guard !isLowPowerModeEnabled else {
            return .lowPowerMode
        }
        guard batteryState == .unplugged else {
            return .batteryCharging
        }
        return if batteryLevel < 0.15 {
            .batteryLevel(.below15)
        } else if batteryLevel < 0.25 {
            .batteryLevel(.below25)
        } else if batteryLevel < 0.4 {
            .batteryLevel(.below40)
        } else if batteryLevel < 0.55 {
            .batteryLevel(.below55)
        } else if batteryLevel < 0.75 {
            .batteryLevel(.below75)
        } else {
            .batteryLevel(.above75)
        }
    }
    
    private func queueStatusByThermalState(_ thermalState: ProcessInfo.ThermalState) -> CameraUploadQueueState {
        switch thermalState {
        case .critical: .thermalState(.critical)
        case .serious: .thermalState(.serious)
        case .fair: .thermalState(.fair)
        case .nominal: .defaultMaximum
        @unknown default: .defaultMaximum
        }
    }
}

enum CameraUploadQueueState: Sendable, Equatable {
    enum BatteryLevel: Sendable, Equatable {
        case above75
        case below75
        case below55
        case below40
        case below25
        case below15
    }
    enum ThermalState: Sendable, Equatable {
        case fair
        case serious
        case critical
    }
    case background
    case lowPowerMode
    case batteryCharging
    case batteryLevel(BatteryLevel)
    case thermalState(ThermalState)
    case defaultMaximum
}

extension CameraUploadQueueState {
    var photoConcurrentCount: Int {
        switch self {
        case .background, .batteryLevel(.below25): 1
        case .lowPowerMode, .batteryLevel(.below40): 2
        case .batteryLevel(.below55), .thermalState(.fair): 3
        case .batteryCharging, .batteryLevel(.above75), .batteryLevel(.below75), .defaultMaximum: 4
        case .batteryLevel(.below15), .thermalState(.critical): 0
        case .thermalState(.serious): 1
        }
    }
    
    var photoPausedReason: CameraUploadMediaTypePausedReason? {
        switch self {
        case .batteryLevel(.below15): .lowBattery
        case .thermalState(.critical): .thermalState(.critical)
        default: nil
        }
    }
    
    var videoConcurrentCount: Int {
        switch self {
        case .background, .lowPowerMode, .batteryCharging, .batteryLevel(.above75), .batteryLevel(.below75), .batteryLevel(.below55), .batteryLevel(.below40), .batteryLevel(.below25), .thermalState(.fair), .defaultMaximum: 1
        case .batteryLevel(.below15), .thermalState(.serious), .thermalState(.critical): 0
        }
    }
    
    var videoPausedReason: CameraUploadMediaTypePausedReason? {
        switch self {
        case .batteryLevel(.below15): .lowBattery
        case .thermalState(.serious): .thermalState(.serious)
        case .thermalState(.critical): .thermalState(.critical)
        default: nil
        }
    }
}

private struct CameraUploadQueueStates {
    let photoUploadState: CameraUploadQueueState
    let videoUploadState: CameraUploadQueueState
}

extension CameraUploadQueueStates {
    var photoConcurrentCount: Int {
        photoUploadState.photoConcurrentCount
    }
    
    var photoPausedReason: CameraUploadMediaTypePausedReason? {
        photoUploadState.photoPausedReason
    }
    
    var videoConcurrentCount: Int {
        photoUploadState.videoConcurrentCount
    }
    
    var videoPausedReason: CameraUploadMediaTypePausedReason? {
        photoUploadState.videoPausedReason
    }
}
